src/main/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributor.java

```diff
@@ -66,12 +69,24 @@ private void processActionType(final ActionType actionType) {
     String actionTypeName = actionType.getName();
     RLock lock = redissonClient.getFairLock(lockPrefix + actionTypeName);
     try {
-      List<Action> actions =
-          actionRepo.findSubmittedOrCancelledByActionTypeName(
-              actionType.getName(), appConfig.getActionDistribution().getRetrievalMax());
-      actions.forEach(this::processAction);
-    } finally {
-      lock.unlock();
+      // Wait for a lock. Automatically unlock after a certain amount of time to prevent issues
+      // when lock holder crashes or Redis crashes causing permanent lockout
+      if (lock.tryLock(
+          appConfig.getDataGrid().getLockTimeToWaitSeconds(),
+          appConfig.getDataGrid().getLockTimeToLiveSeconds(),
+          TimeUnit.SECONDS)) {
+        try {
+          List<Action> actions =
+              actionRepo.findSubmittedOrCancelledByActionTypeName(
```

Comment:
This could actually be improved a lot. At the moment it only grabs a small batch (`retrievalMax`) and it could actually get a stream to _all_ the actions in the states we need - we'd need a new repo method, but it means we could get rid of the native Postgres query and the DB locking

Reply:
:+1: Will look into this after prunes test

Reply:
you gonna have a look at refactoring it now?